src/xmipp/libraries/data/monogenic.cpp

```diff
@@ -0,0 +1,712 @@
+		bool &doNextIteration)
+{
+	resolution = maxRes - count_res*step;
+
+	freq = sampling/resolution;
+	++count_res;
+
+	double Nyquist = 2*sampling;
+	double aux_frequency;
+	int fourier_idx;
+
+	DIGFREQ2FFT_IDX(freq, volsize, fourier_idx);
+
+	FFT_IDX2DIGFREQ(fourier_idx, volsize, aux_frequency);
+
+	freq = aux_frequency;
+
+	if (fourier_idx == last_fourier_idx)
+	{
+		continueIter = true;
+		return;
+	}
+
+	last_fourier_idx = fourier_idx;
+	resolution = sampling/aux_frequency;
+
+
+	if (count_res == 0)
+		last_resolution = resolution;
+
+	if ( ( resolution<Nyquist ))// || (resolution > last_resolution) )
+	{
+		breakIter = true;
+		return;
+	}
+	//	if ( ( freq>0.495))// || (resolution > last_resolution) )
+	//	{
+	//		std::cout << "Nyquist limit reached" << std::endl;
+	//		doNextIteration = false;
+	//		return;
+	//	}
+
+	freqL = sampling/(resolution + step);
+
+	int fourier_idx_2;
+
+	DIGFREQ2FFT_IDX(freqL, volsize, fourier_idx_2);
+
+	if (fourier_idx_2 == fourier_idx)
+	{
+		if (fourier_idx > 0){
+			FFT_IDX2DIGFREQ(fourier_idx - 1, volsize, freqL);
+		}
+		else{
+			freqL = sampling/(resolution + step);
+		}
+	}
+
+}
+
+
+// AMPLITUDEMONOSIG3D_LPF: Estimates the monogenic amplitude of a HPF map, myfftV is the
+// Fourier transform of that map. The monogenic amplitude is "amplitude", and the
+// parameters "fftVRiesz", "fftVRiesz_aux", "VRiesz", are auxiliar variables that will
+// be defined inside the function (they are input to speed up monores algorithm).
+// Freq, freqL and freqH, are the frequency of the HPF and the tails of a raise cosine.
+// Note that once the monogenic amplitude is estimated, it is low pass filtered to obtain
+// a smooth version and avoid rippling. freq_fourier_x, freq_fourier_y, freq_fourier_z,
+// are the accesible Fourier frequencies along each axis.
+void Monogenic::amplitudeMonoSig3D_LPF(const MultidimArray< std::complex<double> > &myfftV,
+		FourierTransformer &transformer_inv,
+		MultidimArray< std::complex<double> > &fftVRiesz,
+		MultidimArray< std::complex<double> > &fftVRiesz_aux, MultidimArray<double> &VRiesz,
+		double freq, double freqH, double freqL, MultidimArray<double> &iu,
+		Matrix1D<double> &freq_fourier_x, Matrix1D<double> &freq_fourier_y,
+		Matrix1D<double> &freq_fourier_z, MultidimArray<double> &amplitude,
+		int count, FileName fnDebug)
+{
+//	FourierTransformer transformer_inv;
+
+	fftVRiesz.initZeros(myfftV);
+	fftVRiesz_aux.initZeros(myfftV);
+	std::complex<double> J(0,1);
+
+	// Filter the input volume and add it to amplitude
+	long n=0;
+	double ideltal=PI/(freq-freqH);
+
+	for(size_t k=0; k<ZSIZE(myfftV); ++k)
+	{
+		for(size_t i=0; i<YSIZE(myfftV); ++i)
+		{
+			for(size_t j=0; j<XSIZE(myfftV); ++j)
+			{
+				double iun=DIRECT_MULTIDIM_ELEM(iu,n);
+				double un=1.0/iun;
+				if (freqH<=un && un<=freq)
+				{
+					DIRECT_MULTIDIM_ELEM(fftVRiesz, n) = DIRECT_MULTIDIM_ELEM(myfftV, n);
+					DIRECT_MULTIDIM_ELEM(fftVRiesz, n) *= 0.5*(1+cos((un-freq)*ideltal));//H;
```

Comment:
If this code takes too long to execute, let me know.

Reply:
It does not take too long. The profiler says that 70% is the Fourier transform. MonoDir uses this function, and there, the performance is important (too slow). But becuase of the botleneck is the fourier transform I guess it cannot be speeded up, right?